identity/clients/registry.go

```diff
@@ -183,9 +184,21 @@ func (r *Registry) Validate(client *ClientRegistration, clientSecret string, red
 		// and has no configured redirect URIs.
 		redirectURIOK := false
 		for _, urlString := range client.RedirectURIs {
-			if urlString == redirectURIString {
-				redirectURIOK = true
-				break
+			if client.ApplicationType == "native" && urlString == "http://localhost" {
+				u, err := url.Parse(redirectURIString)
```

Comment:
Validating that redirectURIString is avlaid URI might be done outside already? We also could move this to the beginning of the function .....

Reply:
Both parses only need to be done when applicationType == "native", since otherwise it must be an exact match. RedirectURIs are already parsed when they are registed, thus those are ensured to be valid. The one provided by the client could be invalid though. The exact reason why it is invalid, should not be required for the error message which is produced later when the redirect_uri is not valid.

Reply:
urlStringParsed, _ := url.Parse(urlString) is still outside the `if client.ApplicationType`.